cmd/snap-update-ns/change.go

```diff
@@ -57,28 +58,180 @@ func (c Change) String() string {
-			return nil, err
+func (c *Change) createInodes(path, kind string, pokeHoles bool) ([]*Change, error) {
+	var err error
+	var changes []*Change
+
+	// In case we need to create something, some constants.
+	const (
+		mode = 0755
+		uid  = 0
+		gid  = 0
+	)
+
+	// If the element doesn't exist we can attempt to create it.  We will
+	// create the parent directory and then the final element relative to it.
+	// The traversed space may be writable so we just try to create things
+	// first.
+
+	// TODO: refactor this, if possible, with inspection and pre-emptive
+	// creation after the current release ships. This should be possible but
+	// will affect tests heavily (churn, not safe before release).
+retry:
+	switch kind {
+	case "":
+		err = secureMkdirAll(path, mode, uid, gid)
+	case "file":
+		err = secureMkfileAll(path, mode, uid, gid)
+	case "symlink":
+		err = secureMkdirAll(path, mode, uid, gid)
+		if err == nil {
+			// Normally symlinks would be created later but calling
+			// c.lowLevelPerform now lets us check if the medium is
+			// writable and take advantage of the code below.
+			err = c.lowLevelPerform()
+		}
+	}
+	if err2, _ := err.(*ReadOnlyFsError); err2 != nil && pokeHoles {
+		// If the writing failed because the underlying filesystem is read-only
+		// we can construct a writable mimic to fix that.
+		changes, err = createWritableMimic(err2.Path)
+		if err != nil {
+			err = fmt.Errorf("cannot create writable mimic over %q: %s", err2.Path, err)
+		} else {
+			// Clear the flag and try again.
+			pokeHoles = false
+			goto retry
+		}
+	}
+	return changes, err
+}
+
+func (c *Change) ensureTarget() ([]*Change, error) {
+	var changes []*Change
+
+	kind, _ := c.Entry.OptStr("x-snapd.kind")
+	path := c.Entry.Dir
+
+	// We use lstat to ensure that we don't follow a symlink in case one was
+	// set up by the snap. Note that at the time this is run, all the snap's
+	// processes are frozen but if the path is a directory controlled by the
+	// user (typically in /home) then we may still race with user processes
+	// that change it.
+	fi, err := osLstat(path)
+
+	if err == nil {
+		// If the element already exists we just need to ensure it is of
+		// the correct type. The desired type depends on the kind of entry
+		// we are working with.
+		switch kind {
+		case "":
+			if !fi.Mode().IsDir() {
+				err = fmt.Errorf("cannot use %q for mounting, not a directory", path)
+			}
+		case "file":
+			if !fi.Mode().IsRegular() {
+				err = fmt.Errorf("cannot use %q for mounting, not a regular file", path)
+			}
+		case "symlink":
+			// When we want to create a symlink we just need the empty
+			// space so anything that is in the way is a problem.
+			err = fmt.Errorf("cannot create symlink in %q, existing file in the way", path)
+		}
+	} else if os.IsNotExist(err) {
+		// No such file or directory, create it
+		if kind == "symlink" {
+			// For symlinks the path should refer to the parent directory.
```

Comment:
Curious that this isn't the case for files. What's the catch?

Reply:
Explained during the call and refactored.